### PR TITLE
Use Arrow's ThreadPool to load record batch

### LINKED ISSUE #6: Use Apache Arrow's Executor to manage thread pool
# Problem

We implemented a naive thread pool in Scanner to perform batch prefetch.  Apache Arrow offers a thread executor for other data formats, we should probably just re-use their thread pool executor, making `Lance` play nicer with the ecosystem.

https://arrow.apache.org/docs/cpp/threading.html
https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/thread_pool.h

# Desired Behavior

Use `Apache Arrow` to manage threads in Lance Reader / Scanner.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #6 and #67.

### CODE REVIEW

**Comment by eddyxu** [comment]
It probably closes #8 as well.

File: cpp/src/lance/io/record_batch_reader.cc [hunk 1]
```diff
@@ -70,19 +80,44 @@ std::shared_ptr<::arrow::Schema> RecordBatchReader::schema() const {
...
-  return f;
+  int total_batches = reader_->metadata().num_batches();
+  while (static_cast<int32_t>(readahead_queue_.size()) < options_->batch_readahead &&
+         current_batch_ < total_batches) {
+    int32_t batch_id = current_batch_++;
+    auto result = thread_pool_->Submit(
+        [&](int32_t batch_id) {
+          /// TODO: how to handle error in thread pool?
+          auto result = this->ReadBatch(batch_id);
+          return result.ValueOrDie();
```
Comment (changhiskhan) [inline cpp/src/lance/io/record_batch_reader.cc:107]: I forgot: does this override an Arrow function? Could we return `Future<Result<shared_ptr<RecordBatch>>>`?

Comment (eddyxu) [reply cpp/src/lance/io/record_batch_reader.cc:107]: The ThreadPool.submit() seems not accept a `Result<T>`, you mean directly return `Future<Result<T>>` within each thread? I can take a look whether it works tho.

File: cpp/src/lance/arrow/file_lance.cc [hunk 2]
```diff
@@ -72,7 +73,8 @@ ::arrow::Result<::arrow::RecordBatchGenerator> LanceFileFormat::ScanBatchesAsync
     offset = lance_fragment_scan_options->offset;
   }
 
-  auto batch_reader = lance::io::RecordBatchReader(reader, options, limit, offset);
+  auto batch_reader = lance::io::RecordBatchReader(
+      reader, options, ::arrow::internal::GetCpuThreadPool(), limit, offset);
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/arrow/file_lance.cc:77]: Previously, in the old Scanner code, we build a threadpool by ourselves. The size of the threadpool was the size of "ScanOptions->batch_readahead". Here, we re-use the thread pool from arrow, which by default, has `# of CPU` threads, as mentioned in #6

File: cpp/src/lance/io/record_batch_reader.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/record_batch_reader.cc]: Yea,  returns `Future<Result<T>>` works. Thanks for the suggestion. 

Updated the code accordingly.

**Review by changhiskhan — APPROVED** [review]
provided build passes

**Comment by eddyxu** [comment]
Filed https://issues.apache.org/jira/browse/ARROW-17309 for arrow to fix manylinux installation bug.

**Pull request merged**